### PR TITLE
WINC-1635, WINC-1592: enable log rotation for kubelet and kubeproxy services

### PR DESCRIPTION
**Pull request by jrvaldes**
This pull request introduces configurable log rotation and flushing for Windows kubelet services, allowing log file size, retention, and flush interval to be set via environment variables. It also adds helper functions to safely parse and use these environment variables, and includes unit tests to ensure their correctness.

Log rotation and flush configuration:

* Added environment variable support for log file size (`SERVICES_LOG_FILE_SIZE`), log file age (`SERVICES_LOG_FILE_AGE`), and log flush interval (`SERVICES_LOG_FLUSH_INTERVAL`), with sensible defaults, to control log runner behavior in `pkg/services/services.go`. [[1]](diffhunk://#diff-c182c95bd24796a29810ef87fd42a281d35041d63c3ba1b6c6eadef9c78430b6R27-R39) [[2]](diffhunk://#diff-ee1cb74eaaa3fdeb736406bb05f9a0438f07328328db086b63b5a4a3b1a30b0bR1-R9)
* Updated the kubelet service command generation to use a new helper function, `getLogRunnerForCmd`, which incorporates the log rotation and flush parameters. [[1]](diffhunk://#diff-c182c95bd24796a29810ef87fd42a281d35041d63c3ba1b6c6eadef9c78430b6L225-R240) [[2]](diffhunk://#diff-c182c95bd24796a29810ef87fd42a281d35041d63c3ba1b6c6eadef9c78430b6R324-R378)
* Added the log flush interval to the kubelet configuration via the `Logging` field, using the new `logsapi.LoggingConfiguration` struct. [[1]](diffhunk://#diff-d5ff1964cc140f787c316cc3b809cd4bf63566860d34d4716fb24f1b171d30b7R755-R760) [[2]](diffhunk://#diff-d5ff1964cc140f787c316cc3b809cd4bf63566860d34d4716fb24f1b171d30b7R25)

Helper functions and environment parsing:

* Introduced `getEnvQuantityOrDefault` and `getEnvDurationOrDefault` functions to safely parse quantity and duration environment variables, falling back to defaults if invalid. [[1]](diffhunk://#diff-c182c95bd24796a29810ef87fd42a281d35041d63c3ba1b6c6eadef9c78430b6R324-R378) [[2]](diffhunk://#diff-ee1cb74eaaa3fdeb736406bb05f9a0438f07328328db086b63b5a4a3b1a30b0bR1-R9)

Testing:

* Added unit tests for log runner command generation and environment variable parsing functions in `pkg/services/services_test.go`, covering various scenarios and edge cases.

## Summary by CodeRabbit

* **New Features**
  * Automatic log rotation for managed Windows services, configurable via environment variables (file size, age, flush interval)
  * Tooling to enable verbose operator debug logging and reduce service log file sizes (applies changes and ensures restart)

* **Updates**
  * Kubelet configuration now includes structured logging with a 5s flush frequency

* **Tests**
  * Added comprehensive tests for env parsing and log-runner command construction

* **Documentation**
  * README updated with log rotation feature and enablement instructions

* **Chores**
  * Dependency declarations adjusted for build consistency

### CODE REVIEW

**Comment by mansikulkarni96** [comment]
okay then do we want to use getLogRunnerForCmd for kube-proxy service configuration? 
If yes, is that too big of a change or can we do it in this PR and another commit?

**Review by mansikulkarni96 — CHANGES_REQUESTED** [review]

File: docs/TROUBLESHOOTING.md [hunk 1]
```diff
@@ -11,6 +11,38 @@ spec:
       value: "--debugLogging"
 ```
 
+### Configuring service log rotation (optional)
```
[the remark this answers is not in the document]
Comment (jrvaldes) [reply docs/TROUBLESHOOTING.md]: logs in the Linux worker nodes are handled by journald

File: docs/TROUBLESHOOTING.md [hunk 2]
```diff
@@ -11,6 +11,38 @@ spec:
...
+    env:
+    - name: SERVICES_LOG_FILE_SIZE
+      value: "100M"  # Rotate when log reaches this size (default: 100M)
+    - name: SERVICES_LOG_FILE_AGE
+      value: "168h"  # Keep rotated logs for this duration (default: 168h/7 days)
+    - name: SERVICES_LOG_FLUSH_INTERVAL
+      value: "5s"    # Flush logs to disk at this interval (default: 5s)
+```
+
+#### Benefits of automatic log rotation:
```
[the remark this answers is not in the document]
Comment (jrvaldes) [reply docs/TROUBLESHOOTING.md]: I feel configuration and defaults should be documented. Do you have a recommendation on where to add them?

**Comment by jrvaldes** [comment]
- https://github.com/openshift/windows-machine-config-operator/pull/3781/ merged, triggering full review

File: pkg/nodeconfig/nodeconfig.go [hunk 3]
```diff
@@ -743,14 +744,20 @@ func generateKubeletConfiguration(clusterDNS string) kubeletconfig.KubeletConfig
...
+		MaxPods:               250,
+		KubeAPIQPS:            &kubeAPIQPS,
+		KubeAPIBurst:          100,
+		SerializeImagePulls:   &falseBool,
+		Logging: logsapi.LoggingConfiguration{
+			FlushFrequency: logsapi.TimeOrMetaDuration{
+				Duration:          meta.Duration{Duration: 5 * time.Second},
+				SerializeAsString: true,
+			},
+		},
```
Comment (sebsoto) [inline pkg/nodeconfig/nodeconfig.go]: This is explicitly setting a [default](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/), which is fine, but the commit message is misleading.

File: pkg/services/services_test.go [hunk 4]
```diff
@@ -154,3 +154,341 @@ func TestHybridOverlayConfiguration(t *testing.T) {
...
+			expectedSuffix := " " + tc.commandPath
+			assert.True(t, len(result) >= len(expectedSuffix) &&
+				result[len(result)-len(expectedSuffix):] == expectedSuffix,
+				"Command path must be at the end of the command string.\nActual: %s", result)
+		})
+	}
+}
+
+// indexOf returns the index of the first occurrence of substr in s, or -1 if not found.
+func indexOf(s, substr string) int {
```
Comment (sebsoto) [inline pkg/services/services_test.go]: https://pkg.go.dev/strings#Index ??

File: pkg/services/init.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package services
+
+import ctrl "sigs.k8s.io/controller-runtime"
+
+var logFileSize, logFileAge, flushInterval string
+
+func init() {
+	log := ctrl.Log.WithName("services").WithName("init")
+
+	var err error
```
Comment (wgahnagl) [inline pkg/services/init.go]: it might feel a little better to save these in a config file somewhere instead of setting them with environment variables? And then address Sebastian's comment about the checking the variables being too complicated by verifying the config all in one go when it loads.

**Comment by jrvaldes** [comment]
> I'm just wondering if setting it through environment variables is the best choice. I think it might be nice to save them in a config somewhere that can get verified.

Makes sense, what do you mean by "can get verified"?

File: hack/common.sh [hunk 6]
```diff
@@ -380,6 +380,10 @@ deleteParallelUpgradeCheckerResources() {
 }
 
 
+# Enables debug logging and set smaller size for services log file in the operator pod to make it easier to
+# troubleshoot issues in CI.
+# The method for patching the deployment depends on the OLM version, which is detected by checking for the presence
+# of a subscription (OLMv0) or clusterextension (OLMv1).
 enable_debug_logging() {
```
[the remark this answers is not in the document]
Comment (jrvaldes) [reply hack/common.sh]: I dont anticipate any issues. @wgahnagl do you recommend creating a specific test to validate this?

File: pkg/services/services.go [hunk 7]
```diff
@@ -307,3 +316,78 @@ func getHostnameCmd(platformType config.PlatformType) string {
 		return ""
 	}
 }
+
+// getLogRunnerForCmd returns the command string to run the given commandPath with kube-log-runner
+// logging to the given logfilePath. Log rotation parameters can be configured via environment variables.
+func getLogRunnerForCmd(commandPath, logfilePath string) string {
+	cmdBuilder := strings.Builder{}
+	// log runner path must be first
```
[the remark this answers is not in the document]
Comment (jrvaldes) [reply pkg/services/services.go]: addressed

**Comment by jrvaldes** [comment]
/override ci/prow/gcp-e2e-operator

passed before

https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_windows-machine-config-operator/3766/pull-ci-openshift-windows-machine-config-operator-master-gcp-e2e-operator/2028946291865686016
